### ISSUE TITLE
explicit_auto_deref gives wrong advise for Boxed types

### ISSUE THREAD

**Issue by asomers** [rust-lang/rust-clippy#9127]
### Summary

The explicit_auto_deref warning suggests a code change that does not compile when trying to get a mutable pointer to the inside of a Box.

Possibly related to #9123

### Lint Name

explicit_auto_deref

### Reproducer

I tried this code:

```rust
struct Foo { pub x: Box<u32> }

pub fn bar(_x: &mut u32) {unimplemented!()}

pub fn baz() {
    let mut foo = Foo{x: Box::new(42u32)};
    bar(&mut *foo.x);
}
```

I saw this happen:

```
warning: deref which would be done by auto-deref
 --> mockall/tests/explicit_auto_deref_box.rs:7:14
  |
7 |     bar(&mut *foo.x);
  |              ^^^^^^ help: try this: `foo.x`
  |
  = note: `#[warn(clippy::explicit_auto_deref)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
```

But when I apply Clippy's advise I get this error:
```
error[E0308]: mismatched types
 --> mockall/tests/explicit_auto_deref_box.rs:7:9
  |
7 |     bar(foo.x);
  |     --- ^^^^^
  |     |   |
  |     |   expected `&mut u32`, found struct `std::boxed::Box`
  |     |   help: consider mutably borrowing here: `&mut foo.x`
  |     arguments to this function are incorrect
  |
  = note: expected mutable reference `&mut u32`
                        found struct `std::boxed::Box<u32>`
```

I expected to see this happen:
No error.

### Version

```text
rustc 1.64.0-nightly (27eb6d701 2022-07-04)
binary: rustc
commit-hash: 27eb6d7018e397cf98d51c205e3576951d766323
commit-date: 2022-07-04
host: x86_64-unknown-freebsd
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by asomers** [comment]
Ahh, I'm stupid.  I'm just so used to seeing this warning flag expressions like `&*xxx` that I didn't read this one closely enough.
